datafusion/common/src/scalar.rs

```diff
@@ -579,24 +581,44 @@ fn dict_from_values<K: ArrowDictionaryKeyType>(
 
 macro_rules! typed_cast_tz {
     ($array:expr, $index:expr, $ARRAYTYPE:ident, $SCALAR:ident, $TZ:expr) => {{
-        let array = $array.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
```

Comment:
typed_cast_tz macro is called by [try_from_array](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L2021-L2022). try_from_array is a public function. Depending on what array value the user passes to the function, downcasting the array to a timestamp array can fail.
This error is reachable and recoverable. It makes sense to return an internal error here. This aligns with how downcast error is handled in the [downcast_value](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/lib.rs#L73) macro.